aip/general/0213.md

```diff
@@ -9,44 +9,82 @@ placement:
 
 # Common components
 
-In general, API reviewers encourage API producers to keep their APIs mostly
-self-contained, except for a relatively small set of common protos which are
-safe to import (e.g. [`google.protobuf.Timestamp`][timestamp]). This is for
-good reason: APIs generally need to be able to move forward independently of
-one another, and mutual dependencies can cause downstream APIs to be forced
-into taking major version changes or even lead to dependency conflicts.
+As specified in [AIP-215][], APIs must be self-contained except for the use
+of "common component" packages which are intended for use by multiple APIs.
 
-However, there are also cases where common structures are valuable, especially
-where a concept is well-known and it is sufficiently clear that it will not
-change. Common protos serve this use case.
+There are two kinds of common component packages:
 
-## Guidance
-
-The public representation of APIs **should** be self-contained, meaning that
-all protos used by the API originate in the same proto `package`, except for
-common protos, which **may** be used freely in any API.
+- Organization-specific common components, covering organization-specific
+  concepts such as a "viewport" in Maps.
+- Global common componnts which are not domain-specific, such as "timestamp" or
+  "postal address".
 
-APIs **must not** define a set of API-specific common protos which live outside
-of its versioning structure. This prevents independent movement of particular
-versions and also causes problems for client libraries in many languages that
-compile the proto messages into classes.
+Where it is safe to share a single representation across multiple APIs,
+common components can make it easier for clients to interact with those APIs.
+Concept-specific client code can be written once, and messages can be used from
+the response of one API in the request of another without clunky copying, for
+example.
 
-APIs **should not** directly depend on protos defined in other APIs. Instead,
-they should copy and paste the applicable messages into their own API. When
-doing so, APIs **should** keep the field names and numbers the same.
+This benefit comes with significant restrictions and limitations, however,
+and should not be attempted lightly.
 
-## Existing common protos
+## Guidance
 
-The common protos, which public-facing protos for an API **may** safely import,
-are as follows:
+- Organization-wide common component packages **must** end with `.type`, e.g.
+  `google.geo.type` or `google.shopping.type`.
+- Organizations **must** consult the API design team before creating a new
+  organization-wide common component package.
+- Organizations **must not** define generic components in organization-specific
+  common component packages, instead preferring global common components.
```

Comment:
I think the previous commit where I commented may have gotten rebased away; not sure what the right way to keep that thread alive is on GitHub.  So I'll just quote my other comment, since I think it still applies:

> ...the way this is currently worded, it sounds like creating `google.actions.type.DateRange` would be forbidden.

Reply:
The intention is that this is covered in the new rationale paragraph:

> Consultation with the API design team is required for global common components
> and suggested for organization-specific common components as the border between
> "generic" and "organization-specific" is a gray area; some generic *concepts*
> have organization-specific use cases which surface through the components.

Reply:
I think the text you quoted sounds good, but A) not sure how discoverable it is; and B) I still think this guidance could be weakened a bit in order not to forbid stuff that is actually sometimes encouraged.

Reply:
I don't have much sympathy for anyone who can't be bothered to read the rationale section :) When viewing the page itself (rather than the source in a diff) it's literally just inches away.

But this isn't forbidding anything that is "actually sometimes encouraged" - it's just that "generic component" has nuance. My understanding is that the guidance section is intended to be short bullet points without too much motivational detail.

I've added an example in the general description just before the guidance section - see if that allays your concerns.

Reply:
It's not just about people not reading the whole thing - it's about the AIPs as citable authorities.  If this text had existed when I made google.actions.type.DateRange, a reviewer could reasonably have cited it in order to prevent the CL from being approved.

Counter-quoting the rationale section ("some generic *concepts* have organization-specific use cases which surface through the components."), or the example above (which says something similar) would have been questionable at best, especially given that I don't think there was anything AoG-specific about google.actions.type.DateRange.  IIRC, the reason not to put it in google.type was just that the Apiness team wanted to see broader usage within google3 before adopting any assumptions about details like date ranges always being inclusive on both ends (e.g. "from 1/1 until 1/3" always including all three days).

How do you feel about the below modification?

```suggestion
- Organizations **must not** define generic components in organization-specific
  common component packages, instead preferring global common components.
  - When there are questions about whether a component qualifies as "generic", please [open an issue][].
```

Anyway, I'm not at Google anymore, and I certainly don't want to block improvements on this based on my opinion on Google's internal process and organization - so if you are still not persuaded then I leave it to you/@toumorokoshi  to figure out the best path forward.  But for the generic AIPs I would probably continue to argue that this guidance, as written, is too strong.

Reply:
I think to clarify: generic AIPs and Google AIPs are not the same thing. So high-level some divergence is allowed. I'd say around "policy" like this especially there's leeway, in contrast to more specific guidance around the schema or behavior of APIs.

> the way this is currently worded, it sounds like creating google.actions.type.DateRange would be forbidden.

I don't really interpret it this way, personally. But we can always change the wording if it's confusing.

I've already approved this PR a while back so I was comfortable with the content then, and my LGTM doesn't change now.

Reply:
Thanks folks.

Creating an issue in this repo would rarely be the right approach - consulting with the API design team internally is. I think "if you want an exception to this, consult with the API design team" is effectively the rule anyway, so I don't think we need to highlight it in this particular example.

Will resolve this conversation now, as @toumorokoshi is okay with how it is. If it turns out that this isn't clear enough, I'm definitely happy to amend it further later.

Reply:
The difficulty I have here is that the term "generic" is not defined at this point. (i.e. it seems approximately synonymous with "common", but that doesn't make the statement meaningful.) So when reading the AIP for the first time, I get stuck at this bullet point because I can't make head or tail of what it means. Yes, the term "generic" is later referenced in the rationale section, but only indirectly. It would help if the term could be defined in the intro section above. I think that would be fairly straightforward.

In the rationale section, it is implied that "generic" is to be seen in contrast to "organization-specific", i.e. as the "other" of the two kinds of common components. So we could modify the bullet point above in line 19 to say something like:

```
- Global common components which are generic, i.e. not domain-specific, such as
  "timestamp" or "postal address".
```

Reply:
Fab - I've put "i.e. not domain-specific" in brackets, but otherwise I've taken your suggestion - thanks!